### PR TITLE
[FEATURE] Greater Flexibility for force update and kill switch

### PR DESCRIPTION
**Pull request by sp71**
- Created Config
- No longer requires kill switch view
- Optional kill switch view or force update view
- Triggers optional delegate methods so views are not necessary
- Updated AppDelegate to a working condition (previously pointing to dead url)
- Inspired from following comments
  - https://github.com/prolificinteractive/Bellerophon/pull/25

### CODE REVIEW

File: Bellerophon/Bellerophon/Sources/BPManager.swift [hunk 1]
```diff
@@ -124,39 +123,48 @@ public class BellerophonManager: NSObject {
...
+        config.allViews().forEach { $0.isHidden = true }
+        view.isHidden = false
+        config.delegate?.bellerophonWillEngage?(self, event: event)
+        bellerophonWindow.makeKeyAndVisible()
+    }
 
-        killSwitchView.frame = killSwitchWindow.bounds
-        delegate?.bellerophonWillEngage?(self)
-        killSwitchWindow.makeKeyAndVisible()
+    internal func view(for event: BellerophonEvent) -> UIView? {
```
Comment (ThibaultKlein) [inline Bellerophon/Bellerophon/Sources/BPManager.swift]: This function would not be necessary if we create an object that combines the event and the view together

File: Bellerophon/Bellerophon/Sources/BPManager.swift [hunk 2]
```diff
@@ -124,39 +123,48 @@ public class BellerophonManager: NSObject {
...
 
     internal func dismissKillSwitchIfNeeded() {
-        guard killSwitchWindow.isKeyWindow else {
+        guard bellerophonWindow.isKeyWindow, let currentEvent = currentEvent else {
             return
         }
-
-        delegate?.bellerophonWillDisengage?(self)
+        config.delegate?.bellerophonWillDisengage?(self, event: currentEvent)
+        config.allViews().forEach { $0.isHidden = true }
```
Comment (ThibaultKlein) [inline Bellerophon/Bellerophon/Sources/BPManager.swift]: Is that necessary if we hide the Bellerophon window? I don't mind adding extra security if you think it's safer

File: Bellerophon/BellerophonTests/BellerophonTests.swift [hunk 3]
```diff
@@ -11,6 +11,11 @@ import XCTest
 
 class BellerophonTests: XCTestCase {
 
+    lazy var mockManager = MockBPManager(config: BellerophonConfig(window: UIWindow(),
```
Comment (ThibaultKlein) [inline Bellerophon/BellerophonTests/BellerophonTests.swift]: Now that you provide a way to pass a force update view or not, can you add tests that cover the expected behavior when `forceUpdateView` is `nil` and when `forceUpdateView` is defined?

File: Example/Bellerophon/AppDelegate.swift [hunk 4]
```diff
@@ -14,29 +14,33 @@ import Bellerophon
 class AppDelegate: UIResponder, UIApplicationDelegate, BellerophonManagerDelegate {
 
     var window: UIWindow?
-    let killSwitchURL = "http://qa.cms.prolific.io/killswitch/status/vandelay/ios"
+    
+    let killSwitchURL = "https://api.myjson.com/bins/19bad7"
+    let forceUpdateURL = "https://api.myjson.com/bins/13f3h7"
+    let perfectApp = "https://api.myjson.com/bins/s3uzf"
```
Comment (ThibaultKlein) [inline Example/Bellerophon/AppDelegate.swift]: Are they hosted using your account? Can you provide some steps on how to update these URLs in case they are not working anymore in the README?

File: Bellerophon/Bellerophon.xcodeproj/xcshareddata/xcschemes/Bellerophon.xcscheme [hunk 5]
```diff
@@ -26,6 +26,7 @@
       buildConfiguration = "Debug"
       selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
       selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
+      language = ""
```
Comment (phillfarrugia) [inline Bellerophon/Bellerophon.xcodeproj/xcshareddata/xcschemes/Bellerophon.xcscheme]: Did this need to change?

File: Bellerophon/Bellerophon/Sources/BPConfig.swift [hunk 6]
```diff
@@ -0,0 +1,54 @@
+//
+//  BellerophonConfig.swift
+//  Bellerophon
+//
+//  Created by Satinder Singh on 11/17/17.
+//
+
+import UIKit
+
+/// Configuration for Bellerophon
```
Comment (phillfarrugia) [inline Bellerophon/Bellerophon/Sources/BPConfig.swift]: This comment is pretty useless. It's obvious from the struct name that it's a config for Bellerophon. I'd perhaps use this comment to describe why this config exists, where it's used and how it's used.

File: Bellerophon/Bellerophon/Sources/BPConfig.swift [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+    /// Initalizer for `BellerophonManager`
+    /// - Note: It is possible to supply no views and still handle force update and killswitch by utilzing methods found in
+    ///         `BellerophonManagerDelegate` via `shouldForceUpdate` and / or `shouldKillSwitch`
+    /// - Parameters:
+    ///   - window: Root UIWindow
+    ///   - killSwitchView: UIView to display when kill switch is triggered. If nil, no view will be shown for kill switch event.
+    ///   - forceUpdateView: UIView to display when force update is triggered. If nil, no view will be shown for force update event.
+    ///   - delegate: `BellerophonManagerDelegate` with optional callbacks.
+    public init(window: UIWindow, killSwitchView: UIView?, forceUpdateView: UIView?, delegate: BellerophonManagerDelegate?) {
```
Comment (phillfarrugia) [inline Bellerophon/Bellerophon/Sources/BPConfig.swift]: default params for killSwitchView, forceUpdateView and delegate so that a client which isn't using them can just call `init(window:)`

File: Bellerophon/Bellerophon/Sources/BPEvent.swift [hunk 8]
```diff
@@ -0,0 +1,18 @@
+//
+//  BPTypes.swift
+//  Bellerophon
+//
+//  Created by Satinder Singh on 11/17/17.
+//
+
+/// Different types of events for Bellerophon
+@objc
+public enum BellerophonEvent: Int {
```
[the remark this answers is not in the document]
Comment (phillfarrugia) [reply Bellerophon/Bellerophon/Sources/BPEvent.swift]: suggestion, perhaps an enum with associated value. So 
```
public enum BellerophonEvent: Int {
  case killSwitch(view: UIView)
  case forceUpdate(view: UIView)
}
```

not sure how compatible this will be with Objc though

File: Bellerophon/Bellerophon/Sources/BPManager.swift [hunk 9]
```diff
@@ -124,39 +123,48 @@ public class BellerophonManager: NSObject {
...
-    }
-
     internal func handleError(error: NSError) {
-        delegate?.receivedError(error: error)
+        config.delegate?.receivedError(error: error)
     }
 
-    internal func displayKillSwitch() {
-        guard !killSwitchWindow.isKeyWindow else {
+    internal func displayWindowIfPossible(for event: BellerophonEvent) {
```
[the remark this answers is not in the document]
Comment (phillfarrugia) [reply Bellerophon/Bellerophon/Sources/BPManager.swift]: agree

File: Bellerophon/Bellerophon/Sources/BPConfig.swift (hunk 7 above)
Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPConfig.swift]: No one would use `Bellerophoron` with `init(window:) ` because it means they are not using any of its functionality. It would keep it clean for the developers if they choose to not use all of them so I suppose I can default nil them all

File: Bellerophon/Bellerophon/Sources/BPEvent.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPEvent.swift]: This suggestion is not compatible with Objective-C. So I'll need to continue making significant changes by now removing `@objc` from the codebase essentially

[the remark this answers is not in the document]
Comment (phillfarrugia) [reply Bellerophon/Bellerophon/Sources/BPEvent.swift]: Suspected so, if not compatible with objc, I'd come up with a different approach. It's definitely do-able while maintaining backwards compatibility.

File: Bellerophon/Bellerophon/Sources/BPManager.swift (hunk 2 above)
Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPManager.swift]: I did it just for safety

File: Bellerophon/BellerophonTests/BellerophonTests.swift (hunk 3 above)
Comment (sp71) [reply Bellerophon/BellerophonTests/BellerophonTests.swift]: will update

File: Bellerophon/Bellerophon.xcodeproj/xcshareddata/xcschemes/Bellerophon.xcscheme (hunk 5 above)
Comment (sp71) [reply Bellerophon/Bellerophon.xcodeproj/xcshareddata/xcschemes/Bellerophon.xcscheme]: I didn't make the change myself. Xcode might have done it since I'm using the latest Xcode

File: Bellerophon/Bellerophon/Sources/BPManager.swift [hunk 10]
```diff
@@ -110,9 +106,16 @@ public class BellerophonManager: NSObject {
...
+                displayWindow(for: .forceUpdate(view: forceUpdateView))
+            }
+            config.delegate?.shouldForceUpdate()
+            startAutoChecking(status)
         } else if status.apiInactive() {
-            displayKillSwitch()
+            if let killSwitchView = config.killSwitchView {
+                displayWindow(for: .killSwitch(view: killSwitchView))
+            }
+            config.delegate?.shouldForceUpdate()
```
Comment (DannyVancura) [inline Bellerophon/Bellerophon/Sources/BPManager.swift]: Should it?

File: Bellerophon/Bellerophon/Sources/BPConfig.swift [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    ///         `BellerophonManagerDelegate` via `shouldForceUpdate` and / or `shouldKillSwitch`
+    /// - Parameters:
+    ///   - window: Root UIWindow
+    ///   - killSwitchView: UIView to display when kill switch is triggered. If nil, no view will be shown for kill switch event.
+    ///   - forceUpdateView: UIView to display when force update is triggered. If nil, no view will be shown for force update event.
+    ///   - delegate: `BellerophonManagerDelegate` with required and optional callbacks.
+    public init(window: UIWindow,
+                killSwitchView: UIView? = nil,
+                forceUpdateView: UIView? = nil,
+                delegate: BellerophonManagerDelegate) {
```
Comment (DannyVancura) [inline Bellerophon/Bellerophon/Sources/BPConfig.swift]: With `delegate` being optional, shouldn't this also be optional and default to nil?

Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPConfig.swift]: There is a required method that triggers the actual network call so the delegate is required

File: Example/Bellerophon/AppDelegate.swift (hunk 4 above)
Comment (sp71) [reply Example/Bellerophon/AppDelegate.swift]: Just updated the README. Great catch

File: Bellerophon/Bellerophon/Sources/BPManagerDelegate.swift [hunk 12]
```diff
@@ -7,42 +7,56 @@
...
     func shouldForceUpdate()
 
-    /// The app is notified when a error was received.
-    ///
-    /// - Parameter error: The error received.
-    func receivedError(error: NSError)
+    /**
+     The app is notified that a kill switch should occur.
+     */
+    func shouldKillSwitch()
```
Comment (ThibaultKlein) [inline Bellerophon/Bellerophon/Sources/BPManagerDelegate.swift]: What is the purpose of this function? Aren't we presenting the kill switch window anyway? I'm assuming this is to cover custom logic when the kill switch is activated. Although, do we have a duplication of behavior with `bellerophonWillEngage`?

File: Bellerophon/Bellerophon/Sources/BPConfig.swift (hunk 6 above)
Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPConfig.swift]: updated

File: Bellerophon/Bellerophon/Sources/BPConfig.swift [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+import UIKit
+
+/// Allows customization settings for Bellerophon.
+/// If killSwitchView is nil, no kill switch view will be displayed.
+/// If forceUpdateView is nil, no force update view will be displayed.
+/// BellerophonManagerDelegate contains required and optional methods
+public struct BellerophonConfig {
+
+    /// The root window
```
[the remark this answers is not in the document]
Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPConfig.swift]: updated

File: Bellerophon/Bellerophon/Sources/BPManagerDelegate.swift (hunk 12 above)
Comment (sp71) [reply Bellerophon/Bellerophon/Sources/BPManagerDelegate.swift]: This method is similar to `shouldForceUpdate`. Developers are no longer required to provide a kill switch view. This gives the option of responding to the kill switch event without having Bellerophon display a kill switch view. `bellerophonWillEngage` will only be called if Bellerophon presents a window to the user. If a developer decides to provide a kill switch view and handle the call back for `shouldKillSwitch`, then it would be duplicated.
